skcosmo/metrics/_reconstruction_measures.py

```diff
@@ -0,0 +1,353 @@
+):
+    # TODO add doc if doc for pointwise_global_reconstruction_error was reviewed
+    (
+        train_size,
+        test_size,
+        random_state,
+        scaler,
+        estimator,
+    ) = _check_reconstruction_measures_input(
+        test_size, train_size, random_state, scaler, estimator
+    )
+
+    X_train, X_test, Y_train, Y_test = train_test_split(
+        X,
+        Y,
+        test_size=test_size,
+        train_size=train_size,
+        random_state=random_state,
+        shuffle=shuffle,
+        train_test_overlap=train_test_overlap,
+    )
+
+    scaler.fit(X_train)
+    X_train = scaler.transform(X_train)
+    X_test = scaler.transform(X_test)
+    scaler.fit(Y_train)
+    Y_train = scaler.transform(Y_train)
+    Y_test = scaler.transform(Y_test)
+
+    squared_dist = (
+        np.sum(X_train ** 2, axis=1)
+        + np.sum(X_test ** 2, axis=1)[:, np.newaxis]
+        - 2 * X_test.dot(X_train.T)
+    )
+
+    n_test = X_test.shape[0]
+
+    def local_reconstruction_error_i(i):
+        local_env_idx = np.argsort(squared_dist[i])[:n_local_points]
+        # D_{k-neigh}^{(i)}
+        local_X_train = X_train[local_env_idx]
+        # \bar{x}_F
+        local_X_train_mean = np.mean(X_train[local_env_idx], axis=0)
+        # D_{k-neigh}^{(i)}
+        local_Y_train = Y_train[local_env_idx]
+        # \bar{x}_F'
+        local_Y_train_mean = np.mean(Y_train[local_env_idx], axis=0)
+        # P_{FF'}
+        estimator.fit(
+            local_X_train - local_X_train_mean,
+            local_Y_train - local_Y_train_mean,
+        )
+        # \tilde{x}_i' = \bar{x}_{F'} + (x_i - \bar{x}_F)P_{FF'}
+        tilde_x_i_dash_test = local_Y_train_mean + estimator.predict(
+            X_test[i, :][np.newaxis, :] - local_X_train_mean
+        )
+        # \|x_i' -  \tilde{x}_i'\|
+        return np.linalg.norm(Y_test[i, :][np.newaxis, :] - tilde_x_i_dash_test)
+
+    pointwise_local_reconstruction_error_values = np.array(
+        Parallel(n_jobs=n_jobs)(
+            delayed(local_reconstruction_error_i)(i) for i in range(n_test)
+        )
+    )
+    return pointwise_local_reconstruction_error_values
+
+
+def local_reconstruction_error(
+    X,
+    Y,
+    n_local_points,
+    test_size=None,
+    train_size=None,
+    random_state=None,
+    shuffle=True,
+    train_test_overlap=False,
+    scaler=None,
+    estimator=None,
+    n_jobs=None,
+):
+    pointwise_local_reconstruction_error_values = pointwise_local_reconstruction_error(
+        X,
+        Y,
+        n_local_points,
+        train_size=train_size,
+        test_size=test_size,
+        random_state=random_state,
+        shuffle=shuffle,
+        train_test_overlap=train_test_overlap,
+        scaler=scaler,
+        estimator=estimator,
+        n_jobs=n_jobs,
+    )
+    return np.linalg.norm(pointwise_local_reconstruction_error_values) / np.sqrt(
+        len(pointwise_local_reconstruction_error_values)
+    )
+
+
+def _check_reconstruction_measures_input(
```

Comment:
Rather than doing it like this, what about defining constants at the top of the script for those which can be re-used and then providing the defaults in the function signatures? This seems unnecessary for something already implemented (default inputs) in python.

Reply:
this is sklearn style, I would say

Reply:
ah now I understand, the reason why defaults are None and is checked is because of what @Luthaf was writing
https://github.com/cosmo-epfl/sklearn-cosmo/pull/9#discussion_r520459829